projects/kbd-2.6.4/docs/man/man8/setkeycodes.8
.\" @(#)man/man8/setkeycodes.8	1.0 Nov  8 22:30:48 MET 1994
.TH SETKEYCODES 8 "8 Nov 1994" "kbd"
.SH NAME
setkeycodes \- load kernel scancode-to-keycode mapping table entries
.SH SYNOPSIS
.B setkeycodes
.I "scancode keycode ..."
.SH DESCRIPTION
The
.I setkeycodes
command reads its arguments two at a time, each pair of arguments
consisting of a scancode (given in hexadecimal) and a keycode (given
in decimal). For each such pair, it tells the kernel keyboard driver
to map the specified scancode to the specified keycode.

This command is useful only for people with slightly unusual keyboards,
that have a few keys which produce scancodes that the kernel does not
recognize.

.SH THEORY
The usual PC keyboard produces a series of scancodes for each
key press and key release. (Scancodes are shown by
\fBshowkey \-s\fP, see
.BR showkey (1)
) The kernel parses this stream of scancodes, and converts it to
a stream of keycodes (key press/release events).
(Keycodes are shown by \fBshowkey\fP.)
Apart from a few scancodes with special meaning, and apart from
the sequence produced by the Pause key, and apart from shiftstate
related scancodes, and apart from the key up/down bit,
the stream of scancodes consists of unescaped
scancodes xx (7 bits) and escaped scancodes e0 xx (8+7 bits).
To these scancodes or scancode pairs, a corresponding keycode can be
assigned (in the range 1-127).
For example, if you have a Macro key that produces e0 6f according
to
.BR showkey (1),
the command
.RS
.B "setkeycodes e06f 112"
.RE
will assign the keycode 112 to it, and then
.BR loadkeys (1)
can be used to define the function of this key.
.LP
Some older kernels might hardwire a low scancode range to the
equivalent keycodes; setkeycodes will fail when you try to remap
these.

.SH "2.6 KERNELS"
In 2.6 kernels key codes lie in the range 1-255, instead of 1-127.
(It might be best to confine oneself to the range 1-239.)
.LP
In 2.6 kernels raw mode, or scancode mode, is not very raw at all.
The code returned by showkey \-s will change after use of setkeycodes.
A kernel bug. See also
.BR showkey (1).
.SH OPTIONS
None.
.SH BUGS
The keycodes of X have nothing to do with those of Linux.
Unusual keys can be made visible under Linux, but not under X.
.SH "SEE ALSO"
.BR dumpkeys (1),
.BR loadkeys (1),
.BR showkey (1),
.BR getkeycodes (8)


projects/kbd-2.6.4/src/setkeycodes.c
/*
 * call: setkeycode scancode keycode ...
 *  (where scancode is either xx or e0xx, given in hexadecimal,
 *   and keycode is given in decimal)
 *
 * aeb, 941108, 2004-01-11
 */
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <sysexits.h>
#include <sys/ioctl.h>
#include <linux/kd.h>

#include "libcommon.h"

static void __attribute__((noreturn))
usage(int rc, const struct kbd_help *options)
{
	const struct kbd_help *h;

	fprintf(stderr, _("Usage: %s [option...] scancode keycode ...\n"), get_progname());
	fprintf(stderr, "\n");
	fprintf(stderr, _("(where scancode is either xx or e0xx, given in hexadecimal,\n"
	                  "and keycode is given in decimal)\n"));

	if (options) {
		int max = 0;

		fprintf(stderr, "\n");
		fprintf(stderr, _("Options:"));
		fprintf(stderr, "\n");

		for (h = options; h && h->opts; h++) {
			int len = (int) strlen(h->opts);
			if (max < len)
				max = len;
		}
		max += 2;

		for (h = options; h && h->opts; h++)
			fprintf(stderr, "  %-*s %s\n", max, h->opts, h->desc);
	}

	fprintf(stderr, "\n");
	fprintf(stderr, _("Report bugs to authors.\n"));
	fprintf(stderr, "\n");

	exit(rc);
}

static int
str_to_uint(const char *str, int base, unsigned int *res)
{
	char *ep;
	long int v;

	errno = 0;
	v = strtol(str, &ep, base);

	if (*ep) {
		kbd_warning(0, _("error reading scancode"));
		return -1;
	}

	if (errno == ERANGE) {
		kbd_warning(0, _("Argument out of range: %s"), str);
		return -1;
	}

	if (v < 0) {
		kbd_warning(0, _("Argument must be positive: %s"), str);
		return -1;
	}

	if (v > UINT_MAX) {
		kbd_warning(0, "Argument is too big: %s", str);
		return -1;
	}

	*res = (unsigned int) v;
	return 0;
}

int main(int argc, char **argv)
{
	int fd, c;
	struct kbkeycode a;
	char *console = NULL;

	set_progname(argv[0]);
	setuplocale();

	const char *short_opts = "C:hV";
	const struct option long_opts[] = {
		{ "console", required_argument, NULL, 'C' },
		{ "help",    no_argument,       NULL, 'h' },
		{ "version", no_argument,       NULL, 'V' },
		{ NULL,      0,                 NULL,  0  }
	};
	const struct kbd_help opthelp[] = {
		{ "-C, --console=DEV", _("the console device to be used.") },
		{ "-V, --version",     _("print version number.")     },
		{ "-h, --help",        _("print this usage message.") },
		{ NULL, NULL }
	};

	while ((c = getopt_long(argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch (c) {
			case 'C':
				if (optarg == NULL || optarg[0] == '\0')
					usage(EX_USAGE, opthelp);
				console = optarg;
				break;
			case 'V':
				print_version_and_exit();
				break;
			case 'h':
				usage(EXIT_SUCCESS, opthelp);
				break;
			case '?':
				usage(EX_USAGE, opthelp);
				break;
		}
	}

	if (optind == argc) {
		kbd_warning(0, _("Not enough arguments."));
		usage(EX_USAGE, opthelp);
	}

	if ((fd = getfd(console)) < 0)
		kbd_error(EX_OSERR, 0, _("Couldn't get a file descriptor referring to the console."));

	while (argc > 2) {
		if (str_to_uint(argv[1], 16, &a.scancode) < 0)
			return EX_DATAERR;

		if (str_to_uint(argv[2], 0, &a.keycode) < 0)
			return EX_DATAERR;

		if (a.scancode >= 0xe000) {
			a.scancode -= 0xe000;
			a.scancode += 128; /* some kernels needed +256 */
		}
#if 0
		/* Test is OK up to 2.5.31--later kernels have more keycodes */
		if (a.scancode > 255 || a.keycode > 127)
			usage(_("code outside bounds"));

		/* Both fields are unsigned int, so can be large;
		   for current kernels the correct test might be
		     (a.scancode > 255 || a.keycode > 239)
		   but we can leave testing to the kernel. */
#endif
		if (ioctl(fd, KDSETKEYCODE, &a)) {
			kbd_error(EXIT_FAILURE, errno,
			          _("failed to set scancode %x to keycode %d: ioctl KDSETKEYCODE"),
			          a.scancode, a.keycode);
		}
		argc -= 2;
		argv += 2;
	}
	return EX_OK;
}
